R/order.R

```diff
@@ -0,0 +1,101 @@
+#' Order and sort vectors
+#'
+#' @description
+#' `vec_order()` computes the order of `x`. For data frames, the order is
+#' computed along the rows by computing the order of the first column and
+#' using subsequent columns to break ties.
+#'
+#' `vec_sort()` sorts `x` by computing its order and using `vec_slice()` to
+#' rearrange.
+#'
+#' `vec_order_loc()` returns a data frame containing a `key` column with
+#' sorted unique groups, and a `loc` column with the locations of each
+#' group in `x`. It is similar to [vec_group_loc()], except the groups are
+#' returned sorted rather than by first appearance.
+#'
+#' @details
+#' Character vectors are ordered in the C-locale. This is different from
+#' `base::order()`, which respects `base::Sys.setlocale()`, but should
+#' produce more reproducible results between different sessions and platforms.
+#' One immediate difference is that ordering is case-sensitive in the C-locale.
+#' Sorting `c("b", "C", "a")` with `vec_sort()` will return `c("C", "a", "b")`,
+#' but with `base::order()` will return `c("a", "b", "C")` unless
+#' `base::order(method = "radix")` is explicitly set, which also uses the
+#' C-locale. That said, this typically does not introduce any practical issues.
+#'
+#' Character vectors are always sorted in UTF-8. If any non-UTF-8
```

Comment:
What does sorted in UTF-8 mean?

Reply:
I should probably write this differently. Character vectors are always converted to UTF-8, and then are sorted one byte at a time. Even though a UTF-8 character can be made of multiple bytes, according to this line in base R's code you can sort them 1 byte at a time (treating them like ASCII) and achieve the same ordering as if you had done it one UTF-8 character at a time. https://github.com/wch/r-source/blob/91b4507bf6040c0167fc5b6037c202c8cbd98afd/src/main/radixsort.c#L967

Reply:
How hard would it be to plug in a proper unicode comparison function? i.e. ignoring how we'd get that function and all the data that it requires to work for different locales, how hard would it be to pass a locale string down, and switch out the existing comparison function?

Reply:
I'm reading a little bit about this. It seems like given some collator C, you can map strings X and Y to their "sort keys" where the following holds:

```
X <= Y
sortkey(X, C) <= sortkey(Y, C)
```

It seems like you are supposed to order sort keys like they are strings themselves, so things like `strcmp()` will order the sort keys in the correct order according to that collation, even if `strcmp()` on X and Y directly doesn't.

I can't say for certain, but it seems like if we can precompute the sort keys, then we can radix sort them using some existing character radix sorting infrastructure (since we are supposed to treat them like strings)?

See "Sort key generation" at the bottom here
http://userguide.icu-project.org/collation

Also this was a useful answer for understanding what the sort key "looks" like
https://stackoverflow.com/questions/12498156/sort-by-collation

Reply:
BTW I think a "sort key" is basically what we call an order proxy.

Reply:
That makes sense.

Another example at the bottom here under “Using large buffers to manage sort keys” that shows how to fill a buffer array with sort keys http://userguide.icu-project.org/collation/examples

Like with strings, each sort key will have variable length so you can’t know the total buffer size up front. But some of the growable vectors ideas used in this PR and elsewhere in vctrs would be useful.

I don’t know how much work it is to check for ICU4C so we could use these functions.